web_client/actions.jsx

```diff
@@ -529,14 +535,16 @@ export const submitLoginForm = (form) => promiseAction(
     let result;
     return (
       rest
-        .login(login, password)
+        .logout()
+        .then(() => rest.login(login, password))
         .then((user) => dispatch(setCurrentUser(user, user.token.token)))
         .then((user) => (result = user))
         .then(() => dispatch(closeDialog()))
         .then(() => result)
         .catch(
           ({ responseJSON: { message } }) => (
             dispatch(setDialogError('login', message))
+              .then(() => dispatch(loginAnonymousUser()))
           )
```

Comment:
I'll elaborate even further on this point:

What we want, *precisely*, is for a catch handler to handle the case where the normal `rest.login()` call fails.  In this case, the app needs to relog as the anonymous user.  The important detail is that we don't want this catch handler to fire in *any* other scenario.

To accomplish this, the promise handler that returns `rest.login(...)` (line 539) should be rewritten to return a promise subchain, with `rest.login(...)` at the head, and the catch handler being the only two links in the chain.

```javascript
return (
  rest
    .logout()
    .then(() => (
      rest.login(login, password)
        .catch(
          ({ responseJSON: { message } }) => (
            ...
          )
        )
    ))
    .then((user) => dispatch(...))
    ...
)
```

Please excuse any typos in the code above.  It has not been tested.